contracts/governance/InterchainMultisig.sol

```diff
@@ -0,0 +1,153 @@
+
+/**
+ * @title InterchainMultisig Contract
+ * @notice Weighted Multisig executor to call functions on any contract
+ */
+contract InterchainMultisig is Caller, BaseWeightedMultisig, IInterchainMultisig {
+    // keccak256('InterchainMultisig.Storage')
+    bytes32 internal constant INTERCHAIN_MULTISIG_STORAGE =
+        0x5a9dc2248a56f285e6221da431581a5990380ebfa07727563571b2be1517a23e;
+
+    using SafeNativeTransfer for address;
+
+    bytes32 public immutable chainNameHash;
+
+    struct InterchainMultisigStorage {
+        mapping(bytes32 => bool) isBatchExecuted;
+    }
+
+    /**
+     * @notice Contract constructor
+     * @dev Sets the initial list of signers and corresponding threshold.
+     * @param chainName The name of the chain
+     * @param weightedSigners The weighted signers payload
+     */
+    constructor(string memory chainName, WeightedSigners memory weightedSigners) BaseWeightedMultisig(0) {
+        chainNameHash = keccak256(bytes(chainName));
+
+        _rotateSigners(weightedSigners);
+    }
+
+    modifier onlySelf() {
+        if (msg.sender != address(this)) revert NotSelf();
+
+        _;
+    }
+
+    /**
+     * @notice Checks if a payload has been executed
+     * @param batchId The hash of the payload payload
+     * @return True if the payload has been executed
+     */
+    function isBatchExecuted(bytes32 batchId) external view returns (bool) {
+        return _interchainMultisigStorage().isBatchExecuted[batchId];
+    }
+
+    /**
+     * @notice Executes an external contract call.
+     * @notice This function is protected by the onlySigners requirement.
+     * @dev Calls a target address with specified calldata and passing provided native value.
+     * @param batchId The batchId of the multisig
+     * @param calls The batch of calls to execute
+     * @param proof The multisig proof data
+     */
+    function executeCalls(
+        bytes32 batchId,
+        Call[] calldata calls,
+        bytes calldata proof
+    ) external payable {
+        InterchainMultisigStorage storage slot = _interchainMultisigStorage();
+        bytes32 messageHash = ECDSA.toEthSignedMessageHash(keccak256(abi.encode(batchId, calls)));
+        uint256 length = calls.length;
+
+        validateProof(messageHash, proof);
+
+        if (slot.isBatchExecuted[batchId]) revert AlreadyExecuted();
+        slot.isBatchExecuted[batchId] = true;
+
+        emit BatchExecuted(batchId, messageHash, length);
+
+        uint256 executedCalls = 0;
+
+        for (uint256 i; i < length; ++i) {
+            Call memory call = calls[i];
+
+            // check if the call is for this contract and chain
+            if (keccak256(bytes(call.chainName)) == chainNameHash && call.executor == address(this)) {
+                if (call.target == address(0)) revert InvalidTarget();
+
+                if (call.target == address(this) && bytes4(call.callData) == InterchainMultisig.voidBatch.selector)
+                    if (length == 1) {
```

Comment:
this doesn't allow voiding for all chains at once. I suggest removing this and renaming `voidBatch` to `noop`, and implicitly let that call be used for voiding. So the logic here just needs to check if at least 1 call for this chain was executed, and revert otherwise. Our scripts can make use of noop for voiding.